### PR TITLE
[rllib] Add microbatch optimizer with A2C example

### PR DESCRIPTION
**Pull request by ericl**
## Why are these changes needed?

This adds a microbatching optimizer example for A2C. The idea is to provide an example for users that may wish to use microbatching with other algorithms.

### CODE REVIEW

File: doc/source/rllib-algorithms.rst [hunk 1]
```diff
@@ -126,6 +126,8 @@ Advantage Actor-Critic (A2C, A3C)
 `[paper] <https://arxiv.org/abs/1602.01783>`__ `[implementation] <https://github.com/ray-project/ray/blob/master/rllib/agents/a3c/a3c.py>`__
 RLlib implements A2C and A3C using SyncSamplesOptimizer and AsyncGradientsOptimizer respectively for policy optimization. These algorithms scale to up to 16-32 worker processes depending on the environment. Both a TensorFlow (LSTM), and PyTorch version are available.
 
+A2C also supports *microbatching*, which can be enabled by setting the ``microbatch_size`` config. Microbatching allows for training with a ``train_batch_size`` much larger than GPU memory. See also the `microbatch optimizer implementation <https://github.com/ray-project/ray/blob/master/rllib/optimizers/microbatch_optimizer.py>`__.
```
Comment (richardliaw) [inline doc/source/rllib-algorithms.rst]: Is microbatching a real word? I've seen "gradient accumulation" everywhere: https://medium.com/huggingface/training-larger-batches-practical-tips-on-1-gpu-multi-gpu-distributed-setups-ec88c3e51255

Comment (ericl) [reply doc/source/rllib-algorithms.rst]: Yeah, see: https://ai.googleblog.com/2019/03/introducing-gpipe-open-source-library.html

Changed docs to reference both.

File: rllib/optimizers/microbatch_optimizer.py [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+
+    @override(PolicyOptimizer)
+    def stats(self):
+        return dict(
+            PolicyOptimizer.stats(self), **{
+                "sample_time_ms": round(1000 * self.sample_timer.mean, 3),
+                "grad_time_ms": round(1000 * self.grad_timer.mean, 3),
+                "update_time_ms": round(1000 * self.update_weights_timer.mean,
+                                        3),
+                "opt_peak_throughput": round(self.grad_timer.mean_throughput,
```
Comment (richardliaw) [inline rllib/optimizers/microbatch_optimizer.py]: peak throughput?

Comment (richardliaw) [reply rllib/optimizers/microbatch_optimizer.py]: isn't this mean throughput?

**Review by richardliaw — APPROVED** [review]

File: rllib/agents/a3c/a2c.py [hunk 3]
```diff
@@ -14,12 +15,31 @@
...
+        # batch of this size until the train batch size is reached. This allows
+        # training with batch sizes much larger than can fit in GPU memory.
+        # To enable, set this to a value less than the train batch size.
+        "microbatch_size": None,
     },
 )
 
+
+def choose_policy_optimizer(workers, config):
+    if config["microbatch_size"]:
```
Comment (richardliaw) [inline rllib/agents/a3c/a2c.py]: should throw an error if micro_batch > train batch size

Comment (ericl) [reply rllib/agents/a3c/a2c.py]: Yep that's done already in the optimizer.
